.github/workflows/test.yml

```diff
@@ -104,10 +104,63 @@ jobs:
         run: |
           rm -f changed_files.txt
 
-  yarn-test:
+  set-base-sha:
+    runs-on: ubuntu-latest
+    outputs:
+      base_sha: ${{ steps.determine-base-sha.outputs.base_sha }}
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v4
+        with:
+          ref: ${{ github.event.pull_request.head.sha || github.sha }}
+
+      - name: Determine BASE_SHA
+        id: determine-base-sha
+        run: |
+          if [ "${{ github.event_name }}" == "pull_request" ]; then
+            echo "base_sha=${{ github.event.pull_request.base.sha }}" >> $GITHUB_OUTPUT
+          elif [ "${{ github.event_name }}" == "merge_group" ]; then
+            echo "base_sha=${{ github.event.merge_group.base_sha }}" >> $GITHUB_OUTPUT
+          elif [ "${{ github.event_name }}" == "push" ]; then
+            echo "base_sha=$${{ github.event.push.before }}" >> $GITHUB_OUTPUT
+          else
+            echo "base_sha=$(git rev-parse HEAD^)" >> $GITHUB_OUTPUT
+          fi
```

Comment:
@antigremlin what do you think we should have as the base case here? just compare github.sha with itself? I don't know how we'll get into a state where CI is running and it's not either a PR, merge queue or push event

Reply:
For a `push` event we could either use `push.before` or rely on `HEAD^`. The real _base_ case here is a manual run which this workflow allows. For a manual run the choice is between comparing `HEAD^` to `HEAD` or not comparing at all, and running the tests for all files.